File: notebooksjava/_04PalindromePermutation.java
import java.util.HashSet;
import java.util.Set;

public class _04PalindromePermutation {

    // 集合の対消し法：奇数回出現する文字が 1 つ以下なら回文に並べ替え可能
    public boolean canPermutePalindrome(String s) {
        s = s.replace(" ", "").toLowerCase();
        Set<Character> oddChars = new HashSet<>();
        for (char c : s.toCharArray()) {
            if (!oddChars.add(c)) {
                oddChars.remove(c);
            }
        }
        return oddChars.size() <= 1;
    }

    public static void main(String[] args) {
        _04PalindromePermutation sol = new _04PalindromePermutation();

        assert sol.canPermutePalindrome("code")    == false : "code → false";
        assert sol.canPermutePalindrome("tactcoa") == true  : "tactcoa → true";
        assert sol.canPermutePalindrome("aab")     == true  : "aab → true";
        assert sol.canPermutePalindrome("a")       == true  : "single char → true";
        assert sol.canPermutePalindrome("")        == true  : "empty string → true";
        assert sol.canPermutePalindrome("Tact Coa") == true : "with spaces/uppercase → true";

        System.out.println("All test cases passed ✅");
    }
}

File: notebooksc++/04-palindrome-permutation.cpp
#include <iostream>
#include <string>
#include <unordered_set>
#include <cctype>
#include <cassert>

using namespace std;

// Using set cancellation method: if at most 1 character appears an odd number of times,
// the string can be rearranged to form a palindrome
bool canPermutePalindrome(string s) {
    // Remove spaces and convert to lowercase
    string cleaned = "";
    for (char c : s) {
        if (c != ' ') {
            cleaned += tolower(c);
        }
    }
    
    unordered_set<char> oddChars;
    
    for (char c : cleaned) {
        if (oddChars.count(c) > 0) {
            oddChars.erase(c);
        } else {
            oddChars.insert(c);
        }
    }
    
    return oddChars.size() <= 1;
}

int main() {
    assert(canPermutePalindrome("code") == false);
    assert(canPermutePalindrome("tactcoa") == true);
    assert(canPermutePalindrome("aab") == true);
    assert(canPermutePalindrome("a") == true);
    assert(canPermutePalindrome("") == true);
    assert(canPermutePalindrome("Tact Coa") == true);
    
    cout << "All test cases passed ✅" << endl;
    return 0;
}
